lib/engine.js

```diff
@@ -38,44 +40,69 @@ class GameEngine {
   step() {
     let state;
 
-    if (!this.lastValidTime) {
-      state = this.initialState;
+    if (this.lastValidTime === null) {
+      const stateJSON = this.initialState;
+
+      state = JSON.parse(stateJSON);
+      this.statesByTime.set(state.time, stateJSON);
+      this.lastValidTime = state.time;
     } else {
-      state = this.statesByTime.get(this.lastValidTime);
+      const stateJSON = this.statesByTime.get(this.lastValidTime);
+
+      if (!stateJSON) {
+        throw new Error('Unexpected cache miss');
+      }
+      state = JSON.parse(stateJSON);
     }
     if (!state) {
       throw new Error('State cache corrupted');
     }
-    state = JSON.parse(state);
-    for (let instant = this.lastValidTime; instant < this.time; ++instant) {
+    for (let instant = this.lastValidTime; instant < this._time; ++instant) {
       const events = this.eventsByTime[instant] || [];
       const effects = this.stepper.step(state, events);
       this.emitEffects(instant, effects);
       this.statesByTime.set(instant + 1, JSON.stringify(state));
     }
-    this.lastValidTime = this.time;
-    this.statesByTime.delete(this.lastValidTime - STATE_CACHE_SIZE);
-    ++this.time;
+    this.lastValidTime = this._time;
+    this.statesByTime.delete(this.lastValidTime - this.stateCacheSize);
+    ++this._time;
 
     this.stepper.postProcess(state);
     return state;
   }
 
   invalidateCache() {
-    this.lastValidTime = 0;
+    this.lastValidTime = null;
     this.statesByTime = new Map();
   }
 
-  addEvent(event) {
-    const events = this.eventsByTime[event.time] || [];
+  get time() {
+    return this._time;
+  }
 
-    if (event.time <= this.lastValidTime - STATE_CACHE_SIZE) {
+  set time(value) {
+    const state = this.statesByTime.get(value);
+
+    if (!state) {
       this.invalidateCache();
     } else {
-      this.lastValidTime = Math.min(this.lastValidTime, event.time);
+      this.lastValidTime = value;
     }
+    this._time = value;
+  }
+
+  addEvent(event) {
+    const events = this.eventsByTime[event.time] || [];
+
     events.push(event);
     this.eventsByTime[event.time] = events;
+    if (this.lastValidTime === null) {
```

Comment:
This could read `lastValidTime === INVALID_TIME` with `const INVALID_TIME = Symbol()`. Remember that `0` is a valid value.